drivers: led: lp5569: add charge-pump configuration + enable-gpios

Emil-Juhl opened the pull request:
[drivers: led: lp5569: add charge pump configuration](https://github.com/zephyrproject-rtos/zephyr/commit/6cbcd23c3811f8065b285c2f2c5f4a26cf4ad288)
The lp5569 controller contains an internal charge pump which can be useful
for driving LEDs with a forward voltage greater than the lp5569 supply.
Taking advantage of the charge pump can alleviate the need for an external
boost converter.

Add a dts property, charge-pump-mode, to the ti,lp5569 binding with which
the cp_mode bits of the MISC register will be configured.

Following the datasheet, the possible configurations are:
    0x00 -> disabled (default)
    0x01 -> 1x mode
    0x10 -> 1.5x mode
    0x11 -> auto mode

[drivers: led: lp5569: add enable-gpios](https://github.com/zephyrproject-rtos/zephyr/commit/74c99cc5c31c2a3eed0da054a3272981bef6f1d5)
The lp5569 features a double functioning pin for enable and pwm control.
The chip, however, uses the first rising edge to initialize and get ready
for i2c communication, and then the pin alters function to pwm mode.

Add support for providing enable-gpios to the lp5569 in the dts, which
will make sure to assert the pin and wait for the chip to initialize
before attempting further configuration of the chip.

Commit:
drivers: led: lp5569: add charge pump configuration

The lp5569 controller contains an internal charge pump which can be useful
for driving LEDs with a forward voltage greater than the lp5569 supply.
Taking advantage of the charge pump can alleviate the need for an external
boost converter.

Add a dts property, charge-pump-mode, to the ti,lp5569 binding with which
the cp_mode bits of the MISC register will be configured.

Following the datasheet, the possible configurations are:
    0x00 -> disabled (default)
    0x01 -> 1x mode
    0x10 -> 1.5x mode
    0x11 -> auto mode

Signed-off-by: Emil Juhl <emdj@bang-olufsen.dk>

Commit:
drivers: led: lp5569: add enable-gpios

The lp5569 features a double functioning pin for enable and pwm control.
The chip, however, uses the first rising edge to initialize and get ready
for i2c communication, and then the pin alters function to pwm mode.

Add support for providing enable-gpios to the lp5569 in the dts, which
will make sure to assert the pin and wait for the chip to initialize
before attempting further configuration of the chip.

Signed-off-by: Emil Juhl <emdj@bang-olufsen.dk>

simonguinot commented on dts/bindings/led/ti,lp5569.yaml:
```diff
@@ -3,3 +3,16 @@ description: TI LP5569 LED
+        0x00 -> disabled (default)
+        0x01 -> 1x
+        0x10 -> 1.5x
+        0x11 -> auto
```
You need to add an `enum` here to specify that this values are OK to take.

Emil-Juhl replied:
Should be taken care of now

simonguinot commented on drivers/led/lp5569.c:
```diff
@@ -156,6 +159,8 @@ static const struct led_driver_api lp5569_led_api = {
 	static const struct lp5569_config lp5569_config_##id = {	\
 		.bus = I2C_DT_SPEC_INST_GET(id),			\
+		.cp_mode = DT_PROP_OR(DT_DRV_INST(id),			\
+				      charge_pump_mode, 0),		\
```
This will ensure that `cp_mode` holds a supported value (which might not be the case with current code).

simonguinot commented on dts/bindings/led/ti,lp5569.yaml:
```diff
@@ -3,3 +3,16 @@ description: TI LP5569 LED
 include: i2c-device.yaml
+
+properties:
+  charge-pump-mode:
```
Also it is worth mentioning this new property in the 4.0 release notes.

Emil-Juhl replied:
Added

simonguinot commented on dts/bindings/led/ti,lp5569.yaml:
```diff
@@ -5,6 +5,11 @@ compatible: "ti,lp5569"
 include: i2c-device.yaml
 
 properties:
+  enable-gpios:
```
I wonder if it would not be better to add a fixed regulator in the devicetree configuration of your board... If not, this addition also needs to go in the release notes.

Emil-Juhl replied:
Since this pin is a dual-purpose enable- and PWM gpio on the chip, that also has some specific timings tied to it, I think it makes sense to describe the pin in the lp5569 nodes.

But functionally, there'd be no issue in tying the enabled part to a regulator with the specific timing described.

simonguinot replied:
You are right.

simonguinot commented on drivers/led/lp5569.c:
```diff
@@ -85,6 +87,25 @@ static int lp5569_enable(const struct device *dev)
+	/* flip the enable pin if specified */
+	if (config->enable_gpio.port) {
+		if (!gpio_is_ready_dt(&config->enable_gpio)) {
+			LOG_ERR("Enable gpio not ready");
```
Suggestion: `GPIO` instead of `gpio`.

simonguinot reviewed (changes requested):
Hi @Emil-Juhl,

Thanks for this PR. Please see my comments below.

Emil-Juhl commented:
> Hi @Emil-Juhl,
> 
> Thanks for this PR. Please see my comments below.

Thanks for the review! I will visit your comments during this week

Emil-Juhl commented:
Hi @simonguinot sorry for not quite making it during last week. I've updated the PR now, so you can take another look when you please and let me know, if other things need to change :)

simonguinot reviewed (dismissed):
It looks good to me. Thanks again for the patch !

simonguinot commented:
Hi @bbilas. Please, can you have a look ?

bbilas commented on drivers/led/lp5569.c:
```diff
@@ -159,6 +180,8 @@ static const struct led_driver_api lp5569_led_api = {
 #define LP5569_DEFINE(id)						\
 	static const struct lp5569_config lp5569_config_##id = {	\
 		.bus = I2C_DT_SPEC_INST_GET(id),			\
+		.enable_gpio = GPIO_DT_SPEC_GET_OR(DT_DRV_INST(id),	\
```
There is `GPIO_DT_SPEC_INST_GET_OR` which can be used without `DT_DRV_INST`.

`GPIO_DT_SPEC_INST_GET_OR(id, enable_gpios, {0})`

Emil-Juhl replied:
Got it

bbilas commented on dts/bindings/led/ti,lp5569.yaml:
```diff
@@ -3,3 +3,16 @@ description: TI LP5569 LED
+      - "1x"
+      - "1x5"
+      - "auto"
+    description: |
```
@simonguinot Shouldn't a default value be noted in the description?

simonguinot replied:
Well, it is the rule according to https://docs.zephyrproject.org/latest/build/dts/bindings-upstream.html#rules-for-default-values. So yes :)

Emil-Juhl replied:
Ah, right. Sorry about that! Should be done now

Commit:
drivers: led: lp5569: run clang-format

Run clang-format on drivers/led/lp5569.c to align formatting in the file.

No functional changes intended.

Signed-off-by: Emil Dahl Juhl <emdj@bang-olufsen.dk>

Emil-Juhl commented:
@simonguinot @bbilas thank you for your reviews!

I believe I've adressed the comments.
Finally I decided to run clang-format on the file to align the formatting to respect the 100 character limit. I don't really know if you guys want this? Alternatively I can redo that commit and only reformat the device macro at the bottom of the file (after requested changes, the shorter lines became a hassle to match).

bbilas commented:
> @simonguinot @bbilas thank you for your reviews!
> 
> I believe I've adressed the comments. Finally I decided to run clang-format on the file to align the formatting to respect the 100 character limit. I don't really know if you guys want this? Alternatively I can redo that commit and only reformat the device macro at the bottom of the file (after requested changes, the shorter lines became a hassle to match).

That is okay for me.

simonguinot commented:
> > @simonguinot @bbilas thank you for your reviews!
> > I believe I've adressed the comments. Finally I decided to run clang-format on the file to align the formatting to respect the 100 character limit. I don't really know if you guys want this? Alternatively I can redo that commit and only reformat the device macro at the bottom of the file (after requested changes, the shorter lines became a hassle to match).
> 
> That is okay for me.

And for me too !

simonguinot reviewed (approved):
Thanks !
